File: Java/66. Plus One.java
/**
 * https://leetcode.com/problems/plus-one/
 */

class Solution {
    // O(n) time complexity
    // O(digits.length + 1) space complexity == O(n) space complexity
    public int[] plusOne(int[] digits) {
        int len = digits.length;
        for (int i = len - 1; i >= 0; i--) {
            if (digits[i] != 9) {
                digits[i] += 1;
                return digits;
            } else {
                digits[i] = 0;
            }
        }
        int[] res = new int[len + 1];
        res[0] = 1;
        return res;
    }
}

File: Kotlin/66. Plus One.kt
/**
 * https://leetcode.com/problems/plus-one/
 */

class Solution {
    // O(n) time complexity
    // O(n) space complexity
    fun plusOne(digits: IntArray): IntArray {
        val len = digits.size
        for (i in len - 1 downTo 0) {
            if (digits[i] == 9) {
                digits[i] = 0
            } else {
                digits[i] += 1
                return digits
            }
        }
        val res = IntArray(len + 1)
        res[0] = 1
        return res
    }
}